TSL: ScriptableNode

sunag opened the pull request:
Dependencies: https://github.com/mrdoob/three.js/pull/25683

**Description**

`ScriptableNode` might seem like a redundant tool out of context, it is a JavaScript API for Nodes and the first release for CPU/GPU Nodes. `ScriptableNode` allows the serialization of scripts and maintains their connections, this opens possibilities for Node Editors who want to use TSL as a shader language.

`ScriptableNode` consider each individual code a unique Node, following this reasoning, was designed to connect with other `ScriptableNode`.

Simple Add for GPU Node
```js
const scriptableNode = js( `

	layout = {
		elements: [
			{ name: 'a', inputType: 'node' },
			{ name: 'b', inputType: 'node' }
		]
	};

	const { add } = TSL;

	function main() {

		const a = parameters.get( 'a' ) || float();
		const b = parameters.get( 'b' ) || float();

		return add( a, b );

	}

` ).scriptable();

scriptableNode.setParameter( 'a', nodeA );
scriptableNode.setParameter( 'b', nodeB );

material = new MeshBasicNodeMaterial();
material.colorNode = scriptableNode;
```

### Visual example
![image](https://user-images.githubusercontent.com/502810/225810616-18a6a8ee-0159-4174-88d2-42bdf1f67069.png)

Mugen87 committed:
PointerLockControls: Refactor class structure. (#25665)

dependabot[bot] committed:
Bump eslint from 8.35.0 to 8.36.0 (#25666)

Bumps [eslint](https://github.com/eslint/eslint) from 8.35.0 to 8.36.0.
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/main/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v8.35.0...v8.36.0)

---
updated-dependencies:
- dependency-name: eslint
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

llzen44 committed:
Update webgl_loader_md2.html (#25670)

Double index increment preventing all animations to be set in UI.

newstart0514 committed:
Fix the Chinese translation of MeshToonMaterial (#25672)

In the original version, the Chinese translation of MeshToonMaterial documents did not work properly. This update aims to fix this error

dependabot[bot] committed:
Bump puppeteer-core from 19.7.4 to 19.7.5 (#25671)

Bumps [puppeteer-core](https://github.com/puppeteer/puppeteer) from 19.7.4 to 19.7.5.
- [Release notes](https://github.com/puppeteer/puppeteer/releases)
- [Changelog](https://github.com/puppeteer/puppeteer/blob/main/release-please-config.json)
- [Commits](https://github.com/puppeteer/puppeteer/compare/puppeteer-core-v19.7.4...puppeteer-core-v19.7.5)

---
updated-dependencies:
- dependency-name: puppeteer-core
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

donmccurdy committed:
Create renovate.json (#25636)

* Create renovate.json

* Update renovate.json

---------

Co-authored-by: mrdoob <info@mrdoob.com>

mrdoob committed:
Delete dependabot.yml

linbingquan committed:
Src: export AnimationAction (#25676)

newstart0514 committed:
Fix a document translation omission (#25677)

* Create a translation document of DataArrayTexture

Create a Chinese translation document of DataArrayTexture

* Update DataArrayTexture.html

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

Mugen87 committed:
Delete DataTexture2DArray.html

WestLangley committed:
CSS2DRenderer: Introduce CSS2DObject.center (#25673)

* CSS2DRenderer: Introduce CSS2DObject.center

* Update css2d_label screenshot

* fix gui

Mugen87 committed:
Vector2: Add `angleTo()`. (#25678)

sunag committed:
CodeNode: Added language and serialize.

sunag committed:
TSL: Renamed inversesqrt -> inverseSqrt, faceforward -> faceForward,

sunag committed:
TSL: Added saturate

sunag committed:
TSL: Different params order if method chaining is used.

sunag committed:
Nodes: Change mix order and added FogNode.mixAssign()

sunag committed:
GLTFLoader: Prevents set data uri in texture.name. (#25682)

* GLTFLoader: Prevents set data uri in texture.name.

* cleanup

* fix maybe optional value

* cleanup

sunag committed:
Node: Ignore private properties on serialization.

sunag committed:
Nodes: Added string, arrayBuffer

sunag committed:
Node: Added .getSerializeChildren()

sunag committed:
Added ScriptableNode

sunag committed:
Added scriptable example and serialization test.

sunag committed:
update style

sunag committed:
cleanup

sunag committed:
cleanup

LeviPesin commented:
I think we should put ScriptableNode, ScriptableValueNode, FunctionNode and CallNode into a separate directory -- maybe call it "code". The idea is that these nodes provide direct low-level access to the language the nodes currently run at (JS, GLSL, WGSL).

LeviPesin commented on examples/jsm/nodes/scriptable/ScriptableNode.js:
```diff
@@ -0,0 +1,427 @@
+
+		this._object = method( ...params );
+
+		const layout = this._object.layout;
```
Maybe we should move this to the editor?

Mugen87 committed:
Object3D: Honor `up` in `toJSON()` and `ObjectLoader`. (#25687)

Mugen87 committed:
WebGLRenderer: Add `onShaderError()`. (#25679)

Mugen87 committed:
Updated builds.

sunag committed:
Code Based Node moved to /code/

sunag commented:
I also want to rename `accessors` to `inputs` but better to do that in another PR.

LeviPesin commented on examples/jsm/nodes/code/CodeNode.js:
```diff
@@ -1,4 +1,4 @@
-import Node, { addNodeClass } from './Node.js';
+import Node, { addNodeClass } from '../core/Node.js';
 import { nodeProxy } from '../shadernode/ShaderNode.js';
```
'../' here should be changed to '../../'.

sunag replied:
Why? doesn't seem to work that way

LeviPesin replied:
I mistakenly thought for some reason that directory level of file has changed 🤦‍♂️

sunag replied:
No problems :)

LeviPesin commented:
> I also want to rename `accessors` to `inputs` but better to do that in another PR.

Maybe `objects` will fit more? Because I think things like BufferNode, TextureNode, ReferenceNode (maybe we should move it there?) are more understandable as "objects" rather than "inputs".

sunag committed:
ScriptableNode: Fix async refresh, added .setLocal(), .getLocal()

sunag committed:
added async scriptable and serialization test

sunag commented:
> Maybe objects will fit more? Because I think things like BufferNode, TextureNode, ReferenceNode (maybe we should move it there?) are more understandable as "objects" rather than "inputs".

Inputs refers to `vars, uniforms, textures, buffers, etc`, while objects is a broader abstraction, but we can certainly have both and make that split.

LeviPesin committed:
3DMLoader: Clean up (#25423)

* 3DMLoader: Clean up

* Update 3DMLoader.js

* Update 3DMLoader.js

* Revert materials comparison change

* Fix `elements`

* Update 3DMLoader.js

* Update 3DMLoader.js

* Update 3DMLoader.js

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

sunag committed:
optional output

sunag committed:
added output example

LeviPesin commented on examples/jsm/nodes/code/ScriptableNode.js:
```diff
@@ -0,0 +1,490 @@
+import * as THREE from 'three';
+import * as Nodes from 'three/nodes';
```
This actually creates a circular reference... Maybe it's better to do something similar to such function used in NodeLoader.

sunag committed:
ScriptableNode: Move editor property to function call if needed.

sunag committed:
ScriptableValueNode: Added support to more CPU nodes.

sunag committed:
ShaderNode: Reduces the use of unnecessary conversions.

sunag committed:
webgpu_materials: Added CPU parameter as node.

sunag committed:
ScriptableNode: Move THREE, TSL to global environment vars.

sunag committed:
webgpu_materials: set globals

LeviPesin commented on examples/jsm/nodes/shadernode/ShaderNode.js:
```diff
@@ -255,6 +255,20 @@ const getAutoTypedConstNode = ( value ) => {
+
+	if ( node.isInputNode !== true || node.value === null ) return;
+
+	const value = node.value;
```
I think we should use `getValueType` function (from NodeUtils) here.

sunag committed:
cleanup

puqeko committed:
Examples: Fix webgl_geometry_spline_editor.html (#25690)

Call render() after detach so that the tooltip disappears when clicking away from the selected helper point, as one would expect.

sunag committed:
TSL: Different params order if method chaining is used. (#25683)

* CodeNode: Added language and serialize.

* TSL: Renamed inversesqrt -> inverseSqrt, faceforward -> faceForward,

* TSL: Added saturate

* TSL: Different params order if method chaining is used.

* Nodes: Change mix order and added FogNode.mixAssign()

sunag committed:
fix conflicting

sunag committed:
cleanup

sunag committed:
Revert "cleanup"

This reverts commit 0cf06951c9016829fff6a244d2e5f5a1c827864f.

sunag committed:
Revert "fix conflicting"

This reverts commit baaa14bb5bfcdfcc29d7b8ec5c64b940ec46d8bb, reversing
changes made to 0923c2ec790d82b8c5be2995f634b49b365b08cd.
